assembly.py

```diff
@@ -278,45 +284,42 @@ def assemble_trinity(
         First trim reads with trimmomatic, rmdup with prinseq,
         and random subsample to no more than 100k reads.
     '''
-    if outReads:
-        subsamp_bam = outReads
+
+    if already_preprocessed_bam:
+        subsamp_bam = inBam
+        read_stats = eval(util.file.slurp_file(preprocessed_bam_stats, noErr=True) or '(0,0,0,0,0,0)')
     else:
         subsamp_bam = util.file.mkstempfname('.subsamp.bam')
+        read_stats = trim_rmdup_subsamp_reads(inBam, clipDb, subsamp_bam, n_reads=n_reads)
 
-    picard_opts = {
-                 'CLIPPING_ATTRIBUTE': tools.picard.SamToFastqTool.illumina_clipping_attribute,
-                 'CLIPPING_ACTION': 'X'
-    }
-
-    read_stats = trim_rmdup_subsamp_reads(inBam, clipDb, subsamp_bam, n_reads=n_reads)
-    subsampfq = list(map(util.file.mkstempfname, ['.subsamp.1.fastq', '.subsamp.2.fastq']))
-    tools.picard.SamToFastqTool().execute(subsamp_bam, subsampfq[0], subsampfq[1], picardOptions=tools.picard.PicardTools.dict_to_picard_opts(picard_opts))
-    try:
-        tools.trinity.TrinityTool().execute(subsampfq[0], subsampfq[1], outFasta, JVMmemory=JVMmemory, threads=threads)
-    except subprocess.CalledProcessError as e:
-        if always_succeed:
-            log.warn("denovo assembly (Trinity) failed to assemble input, emitting empty output instead.")
-            util.file.touch(outFasta)
-        else:
-            raise DenovoAssemblyError(*read_stats)
-    os.unlink(subsampfq[0])
-    os.unlink(subsampfq[1])
-
-    if not outReads:
-        os.unlink(subsamp_bam)
+    with tools.picard.SamToFastqTool().execute_tmp(subsamp_bam, sfx='.subsamp', illuminaClipping=True) as subsampfq:
```

Comment:
I like the pattern of using a context manager for intermediate fastq files from bam precursors. It would be nice to have a (perhaps optional) post-`yield` cleanup of the temp files in the context manager.